pettingzoo/tests/api_test.py

```diff
@@ -113,51 +113,84 @@ def test_rewards_dones(env, agent_0):
+        assert set(env.rewards.keys()) == (set(env.agents)), "agents should not be given a reward if they were done last turn"
+        assert set(env.dones.keys()) == (set(env.agents)), "agents should not be given a done if they were done last turn"
+        assert set(env.infos.keys()) == (set(env.agents)), "agents should not be given an info if they were done last turn"
+        assert set(env.agents).issubset(set(env.possible_agents)), "possible agents should include all agents always"
+        if done:
+            live_agents.remove(agent)
+            has_finished.add(agent)
+        assert env.agents == live_agents, "environment must delete agents as the game continues"
+        if env.env_done:
+            assert has_finished == set(env.possible_agents), "not all agents finished, some were skipped over"
+            break
+
         if isinstance(env.observation_spaces[agent], gym.spaces.Box):
             assert env.observation_spaces[agent].dtype == prev_observe.dtype
         if not env.observation_spaces[agent].contains(prev_observe):
             print("Out of bounds observation: ", prev_observe)
 
         assert env.observation_spaces[agent].contains(prev_observe), "Agent's observation is outside of it's observation space"
         test_observation(prev_observe, observation_0)
-        prev_observe = next_observe
-        if not isinstance(env.infos[agent], dict):
+        if not isinstance(env.infos[env.agent_selection], dict):
             warnings.warn("The info of each agent should be a dict, use {} if you aren't using info")
-        assert env.num_agents == len(env.agents), "env.num_agents is not equal to len(env.agents)"
+
+
 
     env.reset()
     reward_0 = env.rewards[env.agent_selection]
     for agent in env.agent_iter(env.num_agents*2):
-        reward, done, info = env.last()
-        if 'legal_moves' in env.infos[agent] and not done:
+        obs, reward, done, info = env.last()
+        if done:
+            action = None
+        elif 'legal_moves' in env.infos[agent] and not done:
             action = random.choice(env.infos[agent]['legal_moves'])
         else:
             action = env.action_spaces[agent].sample()
-        reward, done, info = env.last()
+        obs, reward, done, info = env.last()
         assert isinstance(done, bool), "Done from last is not True or False"
-        assert reward == env.rewards[agent], "Reward from last() and rewards[agent] do not match"
         assert done == env.dones[agent], "Done from last() and rewards[agent] do not match"
         assert info == env.infos[agent], "Info from last() and infos[agent] do not match"
         float(env.rewards[agent])  # "Rewards for each agent must be convertible to float
         test_reward(reward)
-        observation = env.step(action, observe=False)
+        observation = env.step(action)
         assert observation is None, "step(observe=False) must not return anything"
```

Comment:
Should this assert message be changed?

Reply:
I mean having step equal to something at all is problematic

Reply:
Changed the assert message

?? I thought the idea was that step wouldn't return anything. Implicitly, python functions return None when you don't tell them to return anything.